docs/manual/java/guide/broker/code/docs/javadsl/mb/AnotherServiceImpl.java

```diff
@@ -55,4 +62,83 @@ private void subscribeWithMetadata() {
         //#subscribe-to-topic-with-metadata
 
     }
+
+    private void skipMessages() {
+        //#subscribe-to-topic-skip-messages
+        helloService.greetingsTopic()
+            .subscribe()
+            .atLeastOnce(Flow.fromFunction((GreetingMessage message) -> {
+                if (message.getMessage().equals("Kia ora")) {
+                    return doSomethingWithTheMessage(message);
+                } else {
+                    // Skip all messages where the message is not "Kia ora".
+                    return Done.getInstance();
+                }
+            }));
+        //#subscribe-to-topic-skip-messages
+    }
+
+    //#subscribe-to-topic-partition-messages
+    private void subscribeToHelloMessages() {
+        helloService.greetingsTopic()
+            .subscribe()
+            .atLeastOnce(flow());
+    }
+
+    // This an advanced use of Akka Streams.
+    // See the link above for "working with graphs".
+
+    /**
+     * Creates a `Flow` that partitions the `GreetingMessages` into those that
+     * contain the message "Kia ora" and the others that do not. The ones in
+     * the first partition are sent to the `logic` Flow. The others are sent to
+     * the `skip` Flow where they are mapped directly to `Done`. The Flow's are
+     * then joined back together.
+     *
+     * The resulting flow can be visualized as:
+     * {{{
+     *      +--------------------------------------------------------+
+     *      |  +-------+               +-------+           +------+  |
+     *      |  |       | ~~"Kia ora"~> | logic | ~~Done~~> |      |  |
+     * GreetingMessage |               +-------+           |      |  |
+     *     ~~> |       |                                   | join | ~~> Done
+     *      |  | split |               +-------+           |      |  |
+     *      |  |       |  ~~others~~>  | skip  | ~~Done~~> |      |  |
+     *      |  +-------+               +-------+           +------+  |
+     *      +--------------------------------------------------------+
+     * }}}
+     */
+    private Flow<GreetingMessage, Done, NotUsed> flow() {
+        return Flow.fromGraph(GraphDSL.create(builder -> {
+            final UniformFanOutShape<GreetingMessage, GreetingMessage> split = builder.add(partition());
+            final FlowShape<GreetingMessage, Done> logic = builder.add(logic());
+            final FlowShape<GreetingMessage, Done> ignore = builder.add(skip());
+            final UniformFanInShape<Done, Done> join = builder.add(Merge.create(2));
+            builder.from(split).via(logic).toFanIn(join);
+            builder.from(split).via(ignore).toFanIn(join);
+            return new FlowShape<GreetingMessage, Done>(split.in(), join.out());
+        }));
+    }
```

Comment:
> The only real downside that I can think of is that the complete and failed events do not reach the inner flow. 

There is no issue if the outer flow completes or fails. The inner flow is a new one that is created on each new message, so there is nothing to be propagated downstream. 
And here is the thing, I checked with @ktoso, the downside is more in starting a new stream for each message. 

In any case, I don't think we should bloat the Lagom documentation with this. I think we should only concentrate on how to filter-out some messages like in:

```scala
    helloService
      .greetingsTopic()
      .subscribe
      .atLeastOnce(
        Flow[GreetingMessage].map {
          case msg@GreetingMessage(message) if message.startsWith("Hello") => doSomethingWithTheMessage(msg)
          case _ => Done
        }
      )
```

Could you remove the documentation example on how to use partition and inner flows? I understand that you have spent some time investigating it, but if its a error prone solution, we should not promote it. 

About using a `PartialFunction`, we have this other issue (#844) with some ideas for the Producer API which are also applicable in this case. This could be subject of another PR.

Reply:
@steinybot I'm sorry for the delay in coming back to you. I was traveling the last week and my inbox is really getting behind.

I think the most simple form is enough. The example I posted with `mapAsync` and inner flow is not ideal neither.

Although this solution is not the best one, performance wise, it's better than the original one because it's not error prone. But in any case, I think the gist of it is: how to drop elements and keep offset tracking aligned. Therefore, we should only have the simplest case in the docs. Any other variation of it requires a higher understanding of how streams work and should be not in the Lagom documentation.

To answer your question about the inner/outer flow interactions. The inner flow in my example will never wait for data coming from upstream. The flow is initiated when an element is emitted and it completes once that element is processed. It's lifespan is as long as the each individual `mapAsync` call. In other words, it doesn't stay around awaiting for more elements from the outer flow. That's why it's less than ideal, because we create one flow for each new element.